backend/cmd/headlamp_test.go

```diff
@@ -1339,3 +1340,251 @@ func TestProcessTokenProtocol(t *testing.T) {
 		})
 	}
 }
+
+// setupFakeK8sServer create a mock k8s server for testing purpose,
+// this would help to test Caching Machanism without making request
+// to the k8s server.
+func setupFakeK8sServer(authAllowed bool) *httptest.Server {
+	return httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		if r.URL.Path == "/apis/authorization.k8s.io/v1/selfsubjectaccessreviews" {
+			w.Header().Set("Content-Type", "application/json")
+			w.WriteHeader(http.StatusOK)
+
+			status := fmt.Sprintf(`{"status":{"allowed":%v}}`, authAllowed)
+			_, _ = w.Write([]byte(status))
+
+			return
+		}
+
+		w.Header().Set("Content-Type", "application/json")
+		w.WriteHeader(http.StatusOK)
+
+		if authAllowed {
+			_, _ = w.Write([]byte(`{"kind":"List","apiVersion":"v1","items":[{"metadata":{"name":"resource-test"}}]}`))
+		} else {
+			_, _ = w.Write([]byte(`{{"kind":"Status","apiVersion":"v1","metadata":{"resourceVersion":""},` +
+				`"message":"resource is forbidden: User \"system:serviceaccount:default:test\" cannot get resource ` +
+				`\"resource\" in API group \"\" at the cluster scope","reason":"Forbidden","details":{"kind":"resource"},` +
+				`"code":403}`))
+		}
+	}))
+}
+
+// setupHeadlampConfig create mock HeadlampConfig for testing CacheMiddleware
+// mechanism without creating actual HeadlampConfig.
+func setupHeadlampConfig(fakeK8s *httptest.Server, testName string) *HeadlampConfig {
+	store := kubeconfig.NewContextStore()
+
+	err := store.AddContext(&kubeconfig.Context{
+		ClusterID: fmt.Sprintf("./home/user/kube/config+test-cluster-%s", testName),
+		Name:      "test",
+		Cluster:   &api.Cluster{Server: fakeK8s.URL},
+		AuthInfo:  &api.AuthInfo{Token: "test-token"},
+	})
+	if err != nil {
+		panic(err)
+	}
+
+	return &HeadlampConfig{
+		HeadlampCFG: &headlampconfig.HeadlampCFG{
+			KubeConfigStore: store,
+			CacheEnabled:    true,
+		},
+		telemetryHandler: &telemetry.RequestHandler{},
+		cache:            cache.New[interface{}](),
+	}
+}
+
+// stringResponse converts the response from the request into
+// string value for comparing results.
+func stringResponse(resp *http.Response) (string, error) {
+	bodyBytes, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return "", err
+	}
+
+	bodyString := string(bodyBytes)
+
+	return bodyString, nil
+}
+
+// httpRequestWithContext create request by providing context, url and method, and return
+// http.Response and error.
+func httpRequestWithContext(ctx context.Context, url string, method string) (*http.Response, error) {
+	req, err := http.NewRequestWithContext(ctx, method, url, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	return http.DefaultClient.Do(req)
+}
+
+// TestCacheMiddleware_CacheHitAndCacheMiss test whether the k8s is storing into the cache
+// and returns the data if the data is present in the cache.
+func TestCacheMiddleware_CacheHitAndCacheMiss(t *testing.T) {
+	fakeK8s := setupFakeK8sServer(true)
```

Comment:
Please add this to the top of the integration tests:

```go
	if os.Getenv("HEADLAMP_RUN_INTEGRATION_TESTS") != "true" {
		t.Skip("skipping integration test")
	}
```

This way these tests are not run by default, making the tests a bit faster overall. They are still run if you `HEADLAMP_RUN_INTEGRATION_TESTS=true make backend-test` and they are run in CI.